### PR TITLE
Ignore duplicate records when inserting (fixes #2636)

### LINKED ISSUE #2636: Duplicate entries when sql server has bad connectivity
@Zwelf I guess this is a result of us having multiple master write servers, right? I have to daily fix our replication by skipping inserts that happened on 2 sql servers. Should we just reduce it to 1 sql server for writing?

### PR DESCRIPTION
**Pull request by def-**

### CODE REVIEW

File: scripts/move_sqlite.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def main():
 	print("You can use the following commands to import the entries to MySQL (use sed 's/record_/prefix_/' for other database prefixes):")
 	print()
 	print(("    echo '.dump --preserve-rowids' | sqlite3 {} | " + # including rowids, this forces sqlite to name all columns in each INSERT statement
-			"grep -E '^INSERT INTO record_(race|teamrace|saves)' | " + # filter out inserts
+			"grep -E '^INSERT (IGNORE)? INTO record_(race|teamrace|saves)' | " + # filter out inserts
```
Comment (Zwelf) [inline scripts/move_sqlite.py:89]: The SQLite dump won't include the IGNORE, therefore another line of sed would be necessary to add it to each insert line.

File: src/game/server/score.cpp [hunk 2]
```diff
@@ -508,7 +508,7 @@ bool CScore::SaveScoreThread(IDbConnection *pSqlServer, const ISqlData *pGameDat
 
 	// save score. Can't fail, because no UNIQUE/PRIMARY KEY constrain is defined.
 	str_format(aBuf, sizeof(aBuf),
-			"INSERT INTO %s_race("
+			"INSERT IGNORE INTO %s_race("
```
Comment (Zwelf) [inline src/game/server/score.cpp:511]: I think we have to put a constrain like `PRIMARY KEY (Map, Timestamp, Player)` for this to work (same for record_teamrace).

Comment (def-) [reply src/game/server/score.cpp:511]:
```
MariaDB [teeworlds]> show keys from record_race where Non_unique = 0;;
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| Table       | Non_unique | Key_name | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| record_race |          0 | Map      |            1 | Map         | A         |      119416 |     NULL | NULL   |      | BTREE      |         |               |
| record_race |          0 | Map      |            2 | Name        | A         |     4299003 |     NULL | NULL   |      | BTREE      |         |               |
| record_race |          0 | Map      |            3 | Time        | A         |     4299003 |     NULL | NULL   |      | BTREE      |         |               |
| record_race |          0 | Map      |            4 | Timestamp   | A         |     4299003 |     NULL | NULL   |      | BTREE      |         |               |
| record_race |          0 | Map      |            5 | Server      | A         |     4299003 |     NULL | NULL   |      | BTREE      |         |               |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
5 rows in set (0.001 sec)

ERROR: No query specified

MariaDB [teeworlds]> show keys from record_teamrace where Non_unique = 0;;
+-----------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| Table           | Non_unique | Key_name | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment |
+-----------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| record_teamrace |          0 | PRIMARY  |            1 | Map         | A         |        2279 |     NULL | NULL   |      | BTREE      |         |               |
| record_teamrace |          0 | PRIMARY  |            2 | Name        | A         |      480883 |     NULL | NULL   |      | BTREE      |         |               |
| record_teamrace |          0 | PRIMARY  |            3 | Timestamp   | A         |      480883 |     NULL | NULL   |      | BTREE      |         |               |
| record_teamrace |          0 | PRIMARY  |            4 | Time        | A         |      480883 |     NULL | NULL   |      | BTREE      |         |               |
| record_teamrace |          0 | PRIMARY  |            5 | ID          | A         |      480883 |     NULL | NULL   |      | BTREE      |         |               |
| record_teamrace |          0 | Name     |            1 | Name        | A         |       68697 |     NULL | NULL   |      | BTREE      |         |               |
| record_teamrace |          0 | Name     |            2 | ID          | A         |      480883 |     NULL | NULL   |      | BTREE      |         |               |
+-----------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
7 rows in set (0.001 sec)

ERROR: No query specified

MariaDB [teeworlds]> show keys from record_saves where Non_unique = 0;;
+--------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| Table        | Non_unique | Key_name | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment |
+--------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| record_saves |          0 | Map      |            1 | Map         | A         |        4064 |     NULL | NULL   |      | BTREE      |         |               |
| record_saves |          0 | Map      |            2 | Code        | A         |       48779 |     NULL | NULL   |      | BTREE      |         |               |
+--------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
2 rows in set (0.001 sec)

ERROR: No query specified

MariaDB [teeworlds]> show keys from record_maps where Non_unique = 0;;
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| Table       | Non_unique | Key_name | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| record_maps |          0 | PRIMARY  |            1 | Map         | A         |        1663 |     NULL | NULL   |      | BTREE      |         |               |
| record_maps |          0 | PRIMARY  |            2 | Server      | A         |        1663 |     NULL | NULL   |      | BTREE      |         |               |
| record_maps |          0 | PRIMARY  |            3 | Points      | A         |        1663 |     NULL | NULL   |      | BTREE      |         |               |
| record_maps |          0 | PRIMARY  |            4 | Stars       | A         |        1663 |     NULL | NULL   |      | BTREE      |         |               |
| record_maps |          0 | PRIMARY  |            5 | Mapper      | A         |        1663 |     NULL | NULL   |      | BTREE      |         |               |
| record_maps |          0 | PRIMARY  |            6 | Timestamp   | A         |        1663 |     NULL | NULL   |      | BTREE      |         |               |
| record_maps |          0 | Map      |            1 | Map         | A         |        1663 |     NULL | NULL   |      | BTREE      |         |               |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
7 rows in set (0.000 sec)
```
I have set a few keys in our db.

File: scripts/move_sqlite.py (hunk 1 above)
Comment (def-) [reply scripts/move_sqlite.py:89]: Done

File: src/engine/server/databases/connection.cpp [hunk 3]
```diff
@@ -37,6 +38,7 @@ void IDbConnection::FormatCreateTeamrace(char *aBuf, unsigned int BufferSize, co
 				"ID %s NOT NULL, " // VARBINARY(16) for MySQL and BLOB for SQLite
 				"GameID VARCHAR(64), "
 				"DDNet7 BOOL DEFAULT FALSE"
+				"PRIMARY KEY (Map, Name, Timestamp, Time, ID)"
```
Comment (Zwelf) [inline src/engine/server/databases/connection.cpp:41]: I think for teamrank PRIMARY KEY (ID, Name) would be sufficient, but we can merge as it is now.

Comment (def-) [reply src/engine/server/databases/connection.cpp]: Done.

**Comment by def-** [comment]
I'll submit this now so we can run it on the official servers. I'm getting tired of running after the duplicated records.

**Pull request merged**